### PR TITLE
Implement Number protocol for PyNone

### PR DESCRIPTION
**Pull request by hy-kiera**
Implemented based on #3838

### CODE REVIEW

File: vm/src/builtins/singletons.rs [hunk 1]
```diff
@@ -52,6 +54,13 @@ impl PyNone {
     }
 }
 
+impl AsNumber for PyNone {
+    const AS_NUMBER: PyNumberMethods = PyNumberMethods {
+        boolean: Some(|number, _vm| Ok(self::number_downcast(number).value.is_zero()))
```
Comment (youknowone) [inline vm/src/builtins/singletons.rs:59]: is_zero? the value looking swapped

Comment (hy-kiera) [reply vm/src/builtins/singletons.rs:59]: `non_bool()`, which is of `none_as_number`, returns 0 in cpython `Objects/object.c`
```
static int
none_bool(PyObject *v)
{
    return 0;
}
```

Comment (Snowapril) [reply vm/src/builtins/singletons.rs:59]: return zero means false in this case. As our `bool` returns false as you know, we just need to return what `bool` returns exactly.

Comment (youknowone) [reply vm/src/builtins/singletons.rs:59]: oh, sorry. I was confused with bool. 
as you said, it returns false(0). But this code is not simply returning false. make this line to do return false same to the code you shared.

Comment (hy-kiera) [reply vm/src/builtins/singletons.rs]: Hi. I fixed my code. Could you please review the new commit?

File: vm/src/builtins/singletons.rs [hunk 2]
```diff
@@ -39,7 +41,7 @@ impl Constructor for PyNone {
     }
 }
 
-#[pyimpl(with(Constructor))]
+#[pyimpl(flags(BASETYPE), with(Constructor, AsNumber))]
```
Comment (Snowapril) [inline vm/src/builtins/singletons.rs:44]:
```suggestion
#[pyimpl(with(Constructor, AsNumber))]
```
It seems PyNone does not have `BASETYPE` flags in cpython. Is there any reason for adding it?

**Comment by Snowapril** [comment]
Please run `cargo fmt --all` 😊

**Comment by hy-kiera** [comment]
> Please run `cargo fmt --all` 😊

Okay. I'll run it.

File: vm/src/builtins/singletons.rs [hunk 3]
```diff
@@ -52,6 +52,13 @@ impl PyNone {
     }
 }
 
+impl AsNumber for PyNone {
+    const AS_NUMBER: PyNumberMethods = PyNumberMethods {
+        boolean: Some(|number, _vm| Ok(Self::number_downcast(number).bool())),
```
Comment (youknowone) [inline vm/src/builtins/singletons.rs:57]:
```suggestion
        boolean: Some(|number, _vm| Ok(false)),
```
None is singleton. I am pretty sure you can skip actual object value.

Comment (hy-kiera) [reply vm/src/builtins/singletons.rs:57]: Good idea. BTW, what is the object's functions such as `repr()` and `bool()` used be for?

Comment (Snowapril) [reply vm/src/builtins/singletons.rs:57]: They are exposed as python method by #[pymethod] macro attached. You can call them by below way.
```python
>>>>> None.__bool__()
False
>>>>> None.__repr__()
'None'
```

Comment (youknowone) [reply vm/src/builtins/singletons.rs:57]: And those functions @Snowapril referred are used by facade functions `bool(None)` and `repr(None)`

File: vm/src/builtins/singletons.rs [hunk 4]
```diff
@@ -1,7 +1,7 @@
 use super::{PyType, PyTypeRef};
 use crate::{
-    class::PyClassImpl, convert::ToPyObject, types::Constructor, Context, Py, PyObjectRef,
-    PyPayload, PyResult, VirtualMachine,
+    class::PyClassImpl, convert::ToPyObject, protocol::PyNumberMethods, types::AsNumber,
+    types::Constructor, Context, Py, PyObjectRef, PyPayload, PyResult, VirtualMachine,
```
Comment (youknowone) [inline vm/src/builtins/singletons.rs:4]: you can also group type traits
```rust
types::{AsNumber, Constructor}
```

**Review by qingshi163 — APPROVED** [review]
Looks great

**Pull request merged**